File: dawn_go/xfast_trie.go
package xfasttrie

type XFastTrie struct {
	bits  []bool
	ones  []int
	size  int
}

func Create(w int) *XFastTrie {
	u := 1 << uint(w)
	bits := make([]bool, 2*u-1)
	ones := make([]int, 0)
	return &XFastTrie{bits, ones, 0}
}

func (trie *XFastTrie) Insert(x int) {
	trie.insert(x, 0)
}

func (trie *XFastTrie) insert(x int, i int) {
	if i >= len(trie.bits) {
		trie.size++
		return
	}
	bit := x&(1<<uint(i)) != 0
	index := 1 << uint(len(trie.bits)-1-i)
	trie.bits[index-1] = trie.bits[index-1] || bit
	if bit {
		trie.ones = append(trie.ones, index)
		trie.insert(x, i+1)
	} else {
		trie.insert(x, i+1)
	}
}

func (trie *XFastTrie) Predecessor(x int) int {
	return trie.predecessor(x, 0)
}

func (trie *XFastTrie) predecessor(x int, i int) int {
	if i >= len(trie.bits) {
		return -1
	}
	index := 1 << uint(len(trie.bits)-1-i)
	if trie.bits[index-1] {
		leftChild := 2 * index
		rightChild := leftChild + 1
		if trie.bits[rightChild-1] {
			return trie.predecessor(x, i+1)
		} else {
			return trie.ones[0]
		}
	} else {
		return trie.predecessor(x, i+1)
	}
}

File: dawn_ocaml/lib/xfast_trie.ml
open Base

module XFastTrie = struct
  type t = {
    bits : bool array;
    ones : int list;
    size : int;
  }

  let create w =
    let u = 1 lsl w in
    let bits = Array.make (2 * u - 1) false in
    let ones = [] in
    { bits; ones; size = 0 }

  let insert trie x =
    let rec insert_aux trie x i =
      if i >= Array.length trie.bits then
        { trie with size = trie.size + 1 }
      else
        let bit = x land (1 lsl i) <> 0 in
        let index = (1 lsl (Array.length trie.bits - 1 - i)) in
        trie.bits.(index - 1) <- trie.bits.(index - 1) || bit;
        if bit then
          insert_aux { trie with ones = index :: trie.ones } x (i + 1)
        else
          insert_aux trie x (i + 1)
    in
    insert_aux trie x 0

  let predecessor trie x =
    let rec pred_aux trie x i =
      if i >= Array.length trie.bits then None
      else
        let index = (1 lsl (Array.length trie.bits - 1 - i)) in
        if trie.bits.(index - 1) then
          let left_child = 2 * index in
          let right_child = left_child + 1 in
          if trie.bits.(right_child - 1) then
            pred_aux trie x (i + 1)
          else
            Some (List.hd trie.ones)
        else
          pred_aux trie x (i + 1)
    in
    match pred_aux trie x 0 with
    | None -> None
    | Some index ->
       let rec find_predecessor index =
         if index = 0 then None
         else if List.mem index trie.ones then Some (index - 1)
         else find_predecessor (index / 2)
       in
       find_predecessor (index - 1)
end